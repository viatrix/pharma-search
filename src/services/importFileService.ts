import {
  Connection,
  ConnectionOptions,
  createConnection,
  getConnection,
  DatabaseType,
  Repository
} from 'typeorm';
import 'reflect-metadata';
import { AccessTypes, ApplTypes, Product } from '../db/entity/products';
import { Applicant } from '../db/entity/applicants';
import { DosageForm } from '../db/entity/dosageForms';
import { Ingredient } from '../db/entity/ingredients';
import { ProductIngredient } from '../db/entity/productIngredients';

import { Route } from '../db/entity/routes';
import * as config from '../../config.json';
import 'pg';

export type ProductType = {
  ingredients: string[];
  dosageForm: string;
  route: string;
  tradeName: string;
  applicant: string;
  strength: string[];
  applType: ApplTypes;
  applNo: number;
  productNo: number;
  teCode: string;
  approvalDate: Date;
  rld: boolean;
  rs: boolean;
  accessType: AccessTypes;
  applicantFullName: string;
};

export class ImportFileService {
  private connection: Connection;

  private applicantsRepository: Repository<Applicant>;

  private dosageFormsRepository: Repository<DosageForm>;

  private ingredientsRepository: Repository<Ingredient>;

  private productIngredientsRepository: Repository<ProductIngredient>;

  private productsRepository: Repository<Product>;

  private routesRepository: Repository<Route>;

  public async init () {
    const connectionOptions = {
      type: config.typeorm.type as DatabaseType,
      port: config.typeorm.port,
      host: process.env.TYPEORM_HOST,
      username: process.env.TYPEORM_USERNAME,
      password: process.env.TYPEORM_PASSWORD,
      database: process.env.TYPEORM_DATABASE,
      synchronize: config.typeorm.synchronize,
      logging: config.typeorm.logging,
      entities: config.typeorm.entities,
      migrations: config.typeorm.migrations,
      subscribers: config.typeorm.subscribers
    };
    console.log(connectionOptions);
    if (!this.connection) {
      try {
        this.connection = getConnection();
      } catch (err) {
        if (err.name === 'ConnectionNotFoundError') {
          console.info('Connection did not exist');
        } else {
          throw err;
        }
      }
    }
    if (!this.connection) {
      this.connection = await createConnection(connectionOptions as ConnectionOptions);
      console.info('Created a DB connection');
    }
    this.applicantsRepository = this.connection.getRepository(Applicant);
    this.dosageFormsRepository = this.connection.getRepository(DosageForm);
    this.ingredientsRepository = this.connection.getRepository(Ingredient);
    this.productIngredientsRepository = this.connection.getRepository(ProductIngredient);
    this.productsRepository = this.connection.getRepository(Product);
    this.routesRepository = this.connection.getRepository(Route);
  }

  public async clearDb () {
    await this.productIngredientsRepository.delete({});
    await this.productsRepository.delete({});
    await this.applicantsRepository.delete({});
    await this.dosageFormsRepository.delete({});
    await this.ingredientsRepository.delete({});
    await this.routesRepository.delete({});
  }

  public async parseFile (data: string) {
    const dataWithoutHeaders = data.slice(data.indexOf('\n') + 1);
    const lines = dataWithoutHeaders.split('\n');
    for (const line of lines) {
      const trimmedLine = line.replace('\n', '');
      if (trimmedLine.length) {
        try {
          const product = this.parseLine(trimmedLine);
          await this.saveToDb(product);
        } catch (error) {
          if (error.name === 'InconsistentIngredientsStrength') {
            console.error(`Incorrect list of ingredients and strength in line ${trimmedLine}`);
          } else {
            throw error;
          }
        }
      }
    }
  }

  public parseLine (line: string): ProductType {
    const FEDERAL_NOTE = ' **Federal Register determination that product was not discontinued or withdrawn for safety or efficacy reasons**';
    const lineParts = line.split('~');
    if (lineParts.length !== 14) {
      throw new Error(`Line doesn't contain 14 columns separated by ~: ${line}`);
    }
    const rawIngredients = lineParts[0];
    let ingredients: string[];
    if (rawIngredients.includes(' (')) {
      const basicIngredients = rawIngredients.slice(rawIngredients.indexOf(' (') + 2, rawIngredients.indexOf(')'));
      ingredients = basicIngredients.split(';');
    } else {
      ingredients = rawIngredients.split('; ');
    }
    const dosageForm = lineParts[1].slice(0, lineParts[1].indexOf(';'));
    const route = lineParts[1].slice(lineParts[1].indexOf(';') + 1);
    const tradeName = lineParts[2];
    const applicant = lineParts[3];
    let rawStrength = lineParts[4];
    if (rawStrength.includes(FEDERAL_NOTE)) {
      rawStrength = rawStrength.slice(0, rawStrength.indexOf(FEDERAL_NOTE));
    }
    let strength = rawStrength.split(';');
    let balanced = true;
    for (let i = 0; i < strength.length; i++) {
      if (!this.bracketsBalanced(strength[i])) {
        balanced = false;
      }
    }
    if (!balanced) {
      rawStrength = rawStrength.slice(rawStrength.indexOf(' (') + 2, rawStrength.indexOf(')'));
      strength = rawStrength.split(';');
    }
    if (strength.length !== ingredients.length) {
      const err = new Error(`Incorrect list of ingredients and strength for ${tradeName}`);
      err.name = 'InconsistentIngredientsStrength';
      throw err;
    }
    const applTypeString = lineParts[5];
    let applType: ApplTypes;
    switch (applTypeString) {
      case 'N':
        applType = ApplTypes.ORIGINAL;
        break;
      case 'A':
        applType = ApplTypes.GENERIC;
        break;
      default:
        throw new Error(`Invalid appl_type ${applTypeString}`);
    }
    const applNo = +lineParts[6];
    const productNo = +lineParts[7];
    const teCode = lineParts[8];
    const approvalDate = new Date(lineParts[9]);
    const rld = lineParts[10] === 'Yes';
    const rs = lineParts[11] === 'Yes';
    const accessTypeString = lineParts[12];
    let accessType: AccessTypes;
    switch (accessTypeString) {
      case 'RX':
        accessType = AccessTypes.RX;
        break;
      case 'OTC':
        accessType = AccessTypes.OTC;
        break;
      case 'DISCN':
        accessType = AccessTypes.DISCN;
        break;
      default:
        throw new Error(`Invalid access_type ${accessTypeString}`);
    }
    const applicantFullName = lineParts[13];
    return {
      ingredients,
      dosageForm,
      route,
      tradeName,
      applicant,
      strength,
      applType,
      applNo,
      productNo,
      teCode,
      approvalDate,
      rld,
      rs,
      accessType,
      applicantFullName
    };
  }

  public bracketsBalanced (line: string): boolean {
    let balanced = 0;
    for (let i = 0; i < line.length; i++) {
      switch (line[i]) {
        case '(':
          balanced += 1;
          break;
        case ')':
          balanced -= 1;
      }
    }
    return balanced === 0;
  }

  public async saveToDb (product: ProductType) {
    if (product.ingredients.length !== product.strength.length) {
      throw new Error(`Incorrect list of ingredients and strength for ${product.tradeName}`);
    }
    let applicantEntry = await this.applicantsRepository.findOne(product.applicant);
    if (!applicantEntry) {
      applicantEntry = await this.applicantsRepository.save({
        applicantShortName: product.applicant,
        applicantFullName: product.applicantFullName
      });
    }
    let dosageFormEntry = await this.dosageFormsRepository.findOne({ where: { dosageForm: product.dosageForm } });
    if (!dosageFormEntry) {
      dosageFormEntry = await this.dosageFormsRepository.save({
        dosageForm: product.dosageForm
      });
    }
    let routeEntry = await this.routesRepository.findOne({ where: { route: product.route } });
    if (!routeEntry) {
      routeEntry = await this.routesRepository.save({
        route: product.route
      });
    }
    const productEntry = await this.productsRepository.save({
      tradeName: product.tradeName,
      dosageForm: dosageFormEntry,
      route: routeEntry,
      applicant: applicantEntry,
      applType: product.applType,
      applNo: product.applNo,
      productNo: product.productNo,
      teCode: product.teCode,
      approvalDate: product.approvalDate,
      rld: product.rld,
      rs: product.rs,
      accessType: product.accessType
    });
    for (let i = 0; i < product.ingredients.length; i++) {
      let ingredientEntry = await this.ingredientsRepository.findOne({ where: { ingredient: product.ingredients[i] } });
      if (!ingredientEntry) {
        ingredientEntry = await this.ingredientsRepository.save({ ingredient: product.ingredients[i] });
      }
      await this.productIngredientsRepository.save({
        product: productEntry,
        ingredient: ingredientEntry,
        strength: product.strength[i]
      });
    }
  }
}
