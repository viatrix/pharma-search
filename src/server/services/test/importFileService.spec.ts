import { ImportFileService } from '../importFileService';

let importFileService: ImportFileService;

describe('ParseFile', () => {
  beforeAll(() => {
    importFileService = new ImportFileService();
    importFileService.saveToDb = jest.fn();
  });

  it('should parse file with a header and one line of data', async () => {
    importFileService.parseLine = jest.fn();
    const testData =
      'Ingredient~DF;Route~Trade_Name~Applicant~Strength~Appl_Type~Appl_No~Product_No~TE_Code~Approval_Date~RLD~RS~Type~Applicant_Full_Name\n' +
      'BUDESONIDE~AEROSOL, FOAM;RECTAL~UCERIS~VALEANT PHARMS INTL~2MG/ACTUATION~N~205613~001~~Oct 7, 2014~Yes~Yes~RX~VALEANT PHARMACEUTICALS INTERNATIONAL';
    await importFileService.parseFile(testData);
    expect(importFileService.parseLine).toHaveBeenCalledTimes(1);
    expect(importFileService.parseLine).toBeCalledWith(
      'BUDESONIDE~AEROSOL, FOAM;RECTAL~UCERIS~VALEANT PHARMS INTL~2MG/ACTUATION~N~205613~001~~Oct 7, 2014~Yes~Yes~RX~VALEANT PHARMACEUTICALS INTERNATIONAL'
    );
  });

  it('should parse all lines of a file', async () => {
    importFileService.parseLine = jest.fn();
    const testData =
      'Ingredient~DF;Route~Trade_Name~Applicant~Strength~Appl_Type~Appl_No~Product_No~TE_Code~Approval_Date~RLD~RS~Type~Applicant_Full_Name\n' +
      'BUDESONIDE~AEROSOL, FOAM;RECTAL~UCERIS~VALEANT PHARMS INTL~2MG/ACTUATION~N~205613~001~~Oct 7, 2014~Yes~Yes~RX~VALEANT PHARMACEUTICALS INTERNATIONAL\n' +
      'MINOCYCLINE HYDROCHLORIDE~AEROSOL, FOAM;TOPICAL~AMZEEQ~FOAMIX~EQ 4% BASE~N~212379~001~~Oct 18, 2019~Yes~Yes~RX~FOAMIX PHARMACEUTICALS INC\n' +
      'BETAMETHASONE VALERATE~AEROSOL, FOAM;TOPICAL~BETAMETHASONE VALERATE~PERRIGO UK FINCO~0.12%~A~078337~001~AB~Nov 26, 2012~No~No~RX~PERRIGO UK FINCO LTD PARTNERSHIP\n' +
      'BETAMETHASONE VALERATE~AEROSOL, FOAM;TOPICAL~BETAMETHASONE VALERATE~RICONPHARMA LLC~0.12%~A~207144~001~AB~May 24, 2017~No~No~RX~RICONPHARMA LLC\n' +
      'BETAMETHASONE VALERATE~AEROSOL, FOAM;TOPICAL~BETAMETHASONE VALERATE~TARO~0.12%~A~208204~001~AB~May 24, 2017~No~No~RX~TARO PHARMACEUTICAL INDUSTRIES LTD\n' +
      'CLINDAMYCIN PHOSPHATE~AEROSOL, FOAM;TOPICAL~CLINDAMYCIN PHOSPHATE~PERRIGO UK FINCO~1%~A~090785~001~AT~Mar 31, 2010~No~No~RX~PERRIGO UK FINCO LTD PARTNERSHIP\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~GLENMARK PHARMS LTD~0.05%~A~210809~001~AB1~Feb 15, 2019~No~No~RX~GLENMARK PHARMACEUTICALS LTD\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~GLENMARK PHARMS LTD~0.05%~A~211450~001~AB2~Sep 9, 2019~No~No~RX~GLENMARK PHARMACEUTICALS LTD\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~INGENUS PHARMS LLC~0.05%~A~206805~001~AB1~Jul 31, 2017~No~No~RX~INGENUS PHARMACEUTICALS LLC\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~PERRIGO ISRAEL~0.05%~A~077763~001~AB1~Mar 10, 2008~No~No~RX~PERRIGO ISRAEL PHARMACEUTICALS LTD\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~PERRIGO UK FINCO~0.05%~A~201402~001~AB2~Aug 14, 2012~No~No~RX~PERRIGO UK FINCO LTD PARTNERSHIP\n' +
      'CLOBETASOL PROPIONATE~AEROSOL, FOAM;TOPICAL~CLOBETASOL PROPIONATE~TARO~0.05%~A~208779~001~AB1~Oct 4, 2018~No~No~RX~TARO PHARMACEUTICAL INDUSTRIES LTD\n' +
      'ECONAZOLE NITRATE~AEROSOL, FOAM;TOPICAL~ECOZA~GLENMARK~1%~N~205175~001~~Oct 24, 2013~Yes~Yes~RX~GLENMARK THERAPEUTICS INC USA\n' +
      'BETAMETHASONE DIPROPIONATE; CALCIPOTRIENE~AEROSOL, FOAM;TOPICAL~ENSTILAR~LEO PHARMA AS~0.064%;0.005%~N~207589~001~~Oct 16, 2015~Yes~Yes~RX~LEO PHARMA AS\n' +
      'CLINDAMYCIN PHOSPHATE~AEROSOL, FOAM;TOPICAL~EVOCLIN~MYLAN~1%~N~050801~001~AT~Oct 22, 2004~Yes~Yes~RX~MYLAN PHARMACEUTICALS INC';
    await importFileService.parseFile(testData);
    expect(importFileService.parseLine).toHaveBeenCalledTimes(15);
  });
});

describe('ParseLine', () => {
  beforeAll(() => {
    importFileService = new ImportFileService();
  });

  it('should parse a line of data', async () => {
    const testLine =
      'BETAMETHASONE DIPROPIONATE; CALCIPOTRIENE~AEROSOL, FOAM;TOPICAL~ENSTILAR~LEO PHARMA AS~0.064%;0.005%~N~207589~001~~Oct 16, 2015~Yes~Yes~RX~LEO PHARMA AS';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should parse a line with complex ingredients', async () => {
    const testLine =
      'PANCRELIPASE (AMYLASE;LIPASE;PROTEASE)~CAPSULE, DELAYED RELEASE;ORAL~CREON~ABBVIE~60,000USP UNITS;12,000USP UNITS;38,000USP UNITS~N~020725~002~~Apr 30, 2009~Yes~No~RX~ABBVIE INC';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should parse a line with braces in single ingredient', async () => {
    const testLine =
      'FERRIC HEXACYANOFERRATE(II)~CAPSULE;ORAL~RADIOGARDASE (PRUSSIAN BLUE)~HEYL CHEMISCH~500MG~N~021626~001~~Oct 2, 2003~Yes~Yes~RX~HEYL CHEMISCH PHARMAZEUTISHE FABRIK';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should parse a line with complex strength', async () => {
    const testLine =
      'ARTICAINE HYDROCHLORIDE; EPINEPHRINE BITARTRATE~INJECTABLE;INJECTION~ARTICAINE HYDROCHLORIDE AND EPINEPHRINE BITARTRATE~HOSPIRA~4%;EQ 0.017MG BASE/1.7ML (4%;EQ 0.01MG BASE/ML)~A~079138~001~~Jun 18, 2010~No~No~DISCN~HOSPIRA INC';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should parse a line with complex strength with brackets', async () => {
    const testLine =
      'OLIVE OIL; SOYBEAN OIL~EMULSION;INTRAVENOUS~CLINOLIPID 20%~BAXTER HLTHCARE CORP~16%(160GM/1000ML);4%  (40GM/1000ML)~N~204508~001~~Oct 3, 2013~Yes~Yes~RX~BAXTER HEALTHCARE CORP';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should remove federal note from strength', async () => {
    const testLine =
      'TRAZODONE HYDROCHLORIDE~TABLET, EXTENDED RELEASE;ORAL~OLEPTRO~ANGELINI PHARMA~150MG **Federal Register determination that product was not discontinued or withdrawn for safety or efficacy reasons**~N~022411~001~~Feb 2, 2010~Yes~No~DISCN~ANGELINI PHARMA INC';
    const product = await importFileService.parseLine(testLine);
    expect(product).toMatchSnapshot();
  });

  it('should NOT parse a line of data if number of columns is less than 14', async () => {
    const testLine = 'test string';
    try {
      await importFileService.parseLine(testLine);
    } catch (err) {
      expect(err.message).toEqual(
        "Line doesn't contain 14 columns separated by ~: test string"
      );
    }
  });

  it('should NOT parse file if number of columns is more than 14', async () => {
    const testLine =
      'col1~col2~col3~col4~col5~col6~col7~col8~col9~col10~col11~col12~col13~col14~col15';
    try {
      await importFileService.parseLine(testLine);
    } catch (err) {
      expect(err.message).toEqual(
        "Line doesn't contain 14 columns separated by ~: col1~col2~col3~col4~col5~col6~col7~col8~col9~col10~col11~col12~col13~col14~col15"
      );
    }
  });

  it("should NOT parse file if appl_type is not 'N' or 'A'", async () => {
    const testLine =
      'BUDESONIDE~AEROSOL, FOAM;RECTAL~UCERIS~VALEANT PHARMS INTL~2MG/ACTUATION~wrong_appl_type~205613~001~~Oct 7, 2014~Yes~Yes~RX~VALEANT PHARMACEUTICALS INTERNATIONAL\n';
    try {
      await importFileService.parseLine(testLine);
    } catch (err) {
      expect(err.message).toEqual('Invalid appl_type wrong_appl_type');
    }
  });

  it('should NOT parse file if access type is not valid', async () => {
    const testLine =
      'BUDESONIDE~AEROSOL, FOAM;RECTAL~UCERIS~VALEANT PHARMS INTL~2MG/ACTUATION~N~205613~001~~Oct 7, 2014~Yes~Yes~wrong_access_type~VALEANT PHARMACEUTICALS INTERNATIONAL\n';
    try {
      await importFileService.parseLine(testLine);
    } catch (err) {
      expect(err.message).toEqual('Invalid access_type wrong_access_type');
    }
  });
});
