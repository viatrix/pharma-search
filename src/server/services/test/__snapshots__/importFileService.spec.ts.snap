// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParseLine should parse a line of data 1`] = `
Object {
  "accessType": "RX",
  "applNo": 207589,
  "applType": "ORIGINAL",
  "applicant": "LEO PHARMA AS",
  "applicantFullName": "LEO PHARMA AS",
  "approvalDate": 2015-10-15T21:00:00.000Z,
  "dosageForm": "AEROSOL, FOAM",
  "ingredients": Array [
    "BETAMETHASONE DIPROPIONATE",
    "CALCIPOTRIENE",
  ],
  "productNo": 1,
  "rld": true,
  "route": "TOPICAL",
  "rs": true,
  "strength": Array [
    "0.064%",
    "0.005%",
  ],
  "teCode": "",
  "tradeName": "ENSTILAR",
}
`;

exports[`ParseLine should parse a line with braces in single ingredient 1`] = `
Object {
  "accessType": "RX",
  "applNo": 21626,
  "applType": "ORIGINAL",
  "applicant": "HEYL CHEMISCH",
  "applicantFullName": "HEYL CHEMISCH PHARMAZEUTISHE FABRIK",
  "approvalDate": 2003-10-01T21:00:00.000Z,
  "dosageForm": "CAPSULE",
  "ingredients": Array [
    "FERRIC HEXACYANOFERRATE(II)",
  ],
  "productNo": 1,
  "rld": true,
  "route": "ORAL",
  "rs": true,
  "strength": Array [
    "500MG",
  ],
  "teCode": "",
  "tradeName": "RADIOGARDASE (PRUSSIAN BLUE)",
}
`;

exports[`ParseLine should parse a line with complex ingredients 1`] = `
Object {
  "accessType": "RX",
  "applNo": 20725,
  "applType": "ORIGINAL",
  "applicant": "ABBVIE",
  "applicantFullName": "ABBVIE INC",
  "approvalDate": 2009-04-29T21:00:00.000Z,
  "dosageForm": "CAPSULE, DELAYED RELEASE",
  "ingredients": Array [
    "AMYLASE",
    "LIPASE",
    "PROTEASE",
  ],
  "productNo": 2,
  "rld": true,
  "route": "ORAL",
  "rs": false,
  "strength": Array [
    "60,000USP UNITS",
    "12,000USP UNITS",
    "38,000USP UNITS",
  ],
  "teCode": "",
  "tradeName": "CREON",
}
`;

exports[`ParseLine should parse a line with complex strength 1`] = `
Object {
  "accessType": "DISCN",
  "applNo": 79138,
  "applType": "GENERIC",
  "applicant": "HOSPIRA",
  "applicantFullName": "HOSPIRA INC",
  "approvalDate": 2010-06-17T21:00:00.000Z,
  "dosageForm": "INJECTABLE",
  "ingredients": Array [
    "ARTICAINE HYDROCHLORIDE",
    "EPINEPHRINE BITARTRATE",
  ],
  "productNo": 1,
  "rld": false,
  "route": "INJECTION",
  "rs": false,
  "strength": Array [
    "4%",
    "EQ 0.01MG BASE/ML",
  ],
  "teCode": "",
  "tradeName": "ARTICAINE HYDROCHLORIDE AND EPINEPHRINE BITARTRATE",
}
`;

exports[`ParseLine should parse a line with complex strength with brackets 1`] = `
Object {
  "accessType": "RX",
  "applNo": 204508,
  "applType": "ORIGINAL",
  "applicant": "BAXTER HLTHCARE CORP",
  "applicantFullName": "BAXTER HEALTHCARE CORP",
  "approvalDate": 2013-10-02T21:00:00.000Z,
  "dosageForm": "EMULSION",
  "ingredients": Array [
    "OLIVE OIL",
    "SOYBEAN OIL",
  ],
  "productNo": 1,
  "rld": true,
  "route": "INTRAVENOUS",
  "rs": true,
  "strength": Array [
    "16%(160GM/1000ML)",
    "4%  (40GM/1000ML)",
  ],
  "teCode": "",
  "tradeName": "CLINOLIPID 20%",
}
`;

exports[`ParseLine should remove federal note from strength 1`] = `
Object {
  "accessType": "DISCN",
  "applNo": 22411,
  "applType": "ORIGINAL",
  "applicant": "ANGELINI PHARMA",
  "applicantFullName": "ANGELINI PHARMA INC",
  "approvalDate": 2010-02-01T22:00:00.000Z,
  "dosageForm": "TABLET, EXTENDED RELEASE",
  "ingredients": Array [
    "TRAZODONE HYDROCHLORIDE",
  ],
  "productNo": 1,
  "rld": true,
  "route": "ORAL",
  "rs": false,
  "strength": Array [
    "150MG",
  ],
  "teCode": "",
  "tradeName": "OLEPTRO",
}
`;
